l10n_it_fiscalcode/model/res_partner.py

```diff
@@ -1,31 +1,93 @@
 # -*- coding: utf-8 -*-
 # Copyright 2014 Associazione Odoo Italia (<http://www.odoo-italia.org>)
 # Copyright 2016 Andrea Gallina (Apulia Software)
+# Copyright 2016 Giuliano Lotta
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields, api
+from odoo import models, fields, api, _
+from odoo.exceptions import ValidationError
+
+import logging
+_logger = logging.getLogger(__name__)
+
+try:
+    from stdnum.it.codicefiscale import is_valid
+except ImportError:
+    _logger.warning(
+        'stdnum library not found. '
+        'If you plan to use it, please install the python-stdnum library '
+        'for Fiscal Code validation '
+        'from https://pypi.python.org/pypi/python-stdnum/')
 
 
 class ResPartner(models.Model):
+    """ Extends res.partner to add Italian Fiscal Code
+    """
+    # Private attributes
     _inherit = 'res.partner'
 
+    # Fields declaration
+    fiscalcode = fields.Char(
+        'Fiscal Code', size=16, help="Italian Fiscal Code")
+    is_soletrader = fields.Boolean(
+        string='Sole Trader',
+        default=False,
+        readonly=False,
+        help="Checked if company is a sole trader")
+
     @api.multi
-    def check_fiscalcode(self):
+    @api.onchange('is_company')
+    def onchange_iscompany(self):
+        """ if partner is switched from company to person,
+        is_soletrader is set to False because a simple private citizen
+        may not be a sole trader.
+        A proper warning is displayed.
+        """
         for partner in self:
-            if not partner.fiscalcode:
-                return True
-            elif len(partner.fiscalcode) != 16 and partner.individual:
-                return False
-            else:
-                return True
+            if partner.is_company is False:
```

Comment:
maybe better: if not partner.is_company ... type safe for the scope of this test